File: src/main/java/leetcode/ClimbingStairs.java
package leetcode;

/**
 * You are climbing a staircase. It takes n steps to reach the top.
 *
 * Each time you can either climb 1 or 2 steps. In how many distinct ways can
 * you climb to the top?
 */
public class ClimbingStairs {

    /**
     * Find the number of distinct ways to climb n steps.
     *
     * @param n The number of stairs.
     * @return The number of distinct ways to climb the steps.
     */
    public int climbStairs(final int n) {

        if (n < 1 || n > 45) {
            return 0;
        }

        int[] result = new int[n];

        if (n == 1) {
            return 1;
        } else if (n == 2) {
            return 2;
        } else {
            result[0] = 1;
            result[1] = 2;
            for (int i = 2; i < n; i++) {
                result[i] = result[i - 1] + result[i - 2];
            }

            return result[n - 1];
        }
    }

}

File: python/src/climbing_stairs.py
class Solution:
    """
    You are climbing a staircase. It takes n steps to reach the top.

    Each time you can either climb 1 or 2 steps. In how many distinct ways can you climb to the top?
    """

    def climbStairs(self, n: int) -> int:
        """
        Find the number of distinct ways you can climb to the top.

        :param n: The number of steps to the type.
        :type n: Int.
        :return: The number of distinct ways you can climb to the top.
        :rtype: Int.

        The time complexity is O(N) as we iterate up to n. Faster than 28.31% of solutions.

        The space complexity is O(N) for the array. Less memory than 96.12% of solutions.
        """
        if n < 1 or n > 45:
            return 0

        res = [1, 2]
        for i in range(2, n):
            res.append(res[i - 1] + res[i - 2])

        return res[n - 1]
